### DISCUSSION TITLE
Multiple errors when downloading

### DISCUSSION THREAD

**Discussion by meb427** [open-meteo/open-meteo#447] [General]
Hello,

I installed openmeteo-api on different servers and tried to used either docker or native on Ubuntu 22.04 LTS.
The Api server is running fine.
The first downloads like ECMWF seem working but a lot of errors with ICON or meteofrance.

Example:

[ INFO ] Downloading file http://opendata.dwd.de/weather/nwp/icon/grib/12/w_so/icon_global_icosahedral_soil-level_2023092112_000_27_W_SO.grib2.bz2 (App/Curl.swift:92)
[ DEBUG ] Remapping file ./data/download-icon/temp_4341.grib2 (App/Cdo.swift:172)
ECCODES ERROR   :  pack_bytes: Wrong size (20) for section2Padding. It is 0 bytes long
[ INFO ] Downloading file http://opendata.dwd.de/weather/nwp/icon/grib/12/h_snow/icon_global_icosahedral_single-level_2023092112_000_H_SNOW.grib2.bz2 (App/Curl.swift:92)
[ DEBUG ] Remapping file ./data/download-icon/temp_4341.grib2 (App/Cdo.swift:172)
ECCODES ERROR   :  pack_bytes: Wrong size (20) for section2Padding. It is 0 bytes long
[ INFO ] Downloading file http://opendata.dwd.de/weather/nwp/icon/grib/12/cape_ml/icon_global_icosahedral_single-level_2023092112_000_CAPE_ML.grib2.bz2 (App/Curl.swift:92)
[ DEBUG ] Remapping file ./data/download-icon/temp_4341.grib2 (App/Cdo.swift:172)
ECCODES ERROR   :  pack_bytes: Wrong size (20) for section2Padding. It is 0 bytes long
[ INFO ] Downloading file http://opendata.dwd.de/weather/nwp/icon/grib/12/hzerocl/icon_global_icosahedral_single-level_2023092112_000_HZEROCL.grib2.bz2 (App/Curl.swift:92)
[ DEBUG ] Remapping file ./data/download-icon/temp_4341.grib2 (App/Cdo.swift:172)
ECCODES ERROR   :  pack_bytes: Wrong size (20) for section2Padding. It is 0 bytes long
[ INFO ] Downloading file http://opendata.dwd.de/weather/nwp/icon/grib/12/relhum_2m/icon_global_icosahedral_single-level_2023092112_000_RELHUM_2M.grib2.bz2 (App/Curl.swift:92)
[ DEBUG ] Remapping file ./data/download-icon/temp_4341.grib2 (App/Cdo.swift:172)
ECCODES ERROR   :  pack_bytes: Wrong size (20) for section2Padding. It is 0 bytes long
[ INFO ] Downloading hour 1 (App/DownloadIconCommand.swift:129)
[ INFO ] Downloading file http://opendata.dwd.de/weather/nwp/icon/grib/12/t_2m/icon_global_icosahedral_single-level_2023092112_001_T_2M.grib2.bz2 (App/Curl.swift:92)

When I check manually the same link from the same machine. I was able to fetch the file no issues!

I even resorted to set chmod 777 -R on /var/lib/openmeteo-api. 

I'm running the cron under my user should that be under root or openmeteo user?

should I set the cron manually? Or the service will run it? ( I'm a bit lost in this topic.)

Why did you choose to separate surface for icon instead of pulling all icon once for all?

Second issue
Also when I tried to run the cron as is. A lot of zombie process were spawned after 2 days as they endup doing a lot of errors.

May be the cront should use the flock:

flock -n tmp/icon.lock -c "cd /var/lib/openmeteo-api && /usr/local/bin/openmeteo-api download icon --group surface > log/icon.log 2>&1"
 I can do a pull request to update it that way if you are ok.
 
Same error on HRR download:

[ INFO ] Downloading file https://nomads.ncep.noaa.gov/pub/data/nccf/com/hrrr/prod/hrrr.20230921/conus/hrrr.t15z.wrfprsf07.grib2.idx (App/Curl.swift:92)
[ INFO ] Last modified header is too jung. Target delay 120.0 seconds. Sleeping for 112.0 seconds now. (App/Curl.swift:394)
[ INFO ] Downloading [ranged] file https://nomads.ncep.noaa.gov/pub/data/nccf/com/hrrr/prod/hrrr.20230921/conus/hrrr.t15z.wrfprsf07.grib2 (App/Curl.swift:92)
[ INFO ] Downloading forecastHour 8 (App/GfsDownload.swift:197)
[ INFO ] Downloading file https://nomads.ncep.noaa.gov/pub/data/nccf/com/hrrr/prod/hrrr.20230921/conus/hrrr.t15z.wrfprsf08.grib2.idx (App/Curl.swift:92)
[ INFO ] Last modified header is too jung. Target delay 120.0 seconds. Sleeping for 102.0 seconds now. (App/Curl.swift:394)
[ INFO ] Downloading [ranged] file https://nomads.ncep.noaa.gov/pub/data/nccf/com/hrrr/prod/hrrr.20230921/conus/hrrr.t15z.wrfprsf08.grib2 (App/Curl.swift:92)
[ INFO ] Downloading forecastHour 9 (App/GfsDownload.swift:197)
[ INFO ] Downloading file https://nomads.ncep.noaa.gov/pub/data/nccf/com/hrrr/prod/hrrr.20230921/conus/hrrr.t15z.wrfprsf09.grib2.idx (App/Curl.swift:92)
[ INFO ] Download failed, retry every 5 seconds, (0 minutes elapsed, curl error 'downloadFailed(code: 404 Not Found)' (App/TimeoutTracker.swif

While trying same the URL in browser same machine, no issue.

Testing made on Atom N100 either Ubuntu 22.04 native packages and using docker and also on Ryzen 2 under WSL 22.04 installed as packages.

Thanks for your help and feeback.

**Comment by patrick-zippenfenig** [comment]
Hi,

The ECCODES messages `ECCODES ERROR : pack_bytes: Wrong size (20) for section2Padding. It is 0 bytes long` are caused by the grib decoding library. It is still working fine. There is no way to suppress them.

> Why did you choose to separate surface for icon instead of pulling all icon once for all?

Because downloading takes 3 or more hours. Most users also do not use pressure or model level data. High CPU resources are also required to reproject the ICON grid.

> Also when I tried to run the cron as is. A lot of zombie process were spawned after 2 days as they endup doing a lot of errors.

This is not happening on my systems. There are timeouts that usually terminate the process after a couple of hours. Using flock would prevent the next weather model update from downloading. Some weather models (e.g. HRRR) update every hours and may even overlap and multiple processes download different weather model runs at the same time.

> [ INFO ] Last modified header is too jung. Target delay 120.0 seconds. Sleeping for 112.0 seconds now. (App/Curl.swift:394)

`404` and retry errors are expected during download. The downloaders are designed to be started before all weather model data is available and wait until the weather model completes. It may take up to 2 hour from the forecast hour being calculated until the last forecast step is available.

> Testing made on Atom N100 either Ubuntu 22.04 native packages and using docker and also on Ryzen 2 under WSL 22.04 installed as packages.

The downloaders are CPU demanding. A lot of data needs to be decoded, reprojected, supplementary solar position models run and all data compressed again. If you select too many weather variables, the N100 won't be able to convert data in time. I have also no experiences running under WSL. This could explain the zombie processes

**Comment by meb427** [comment]
The zombie processes were on Ubuntu 22.04 on Atom. I think there was a lot of racing among cron, mostly slowed down by HDD used instead of SSD, as I attached a HDD to this PC to get more ERA5 data.

Another infos I noticed when doing the benchmark (on WSL2/Win11 here) but backed with Ryzen 9.

Result

Open-Meteo Benchmark
Apple M1 is used as baseline. Positive values = slower than M1.
Time per test 5 seconds (See --help)
| Test                                                                             | Mean     | Min      | Max      | Runs     | Diff to Apple M1     |
|----------------------------------------------------------------------------------|----------|----------|----------|----------|----------------------|
| Generating dummy temperature timeseries (128 MB)                                 | 682ms    | 671ms    | 693ms    | 8        | +410ms (x2.51)       |
| Compression in memory, large chunks                                              | 133ms    | 131ms    | 139ms    | 38       | -58.3ms (x0.69)      |
| Compression in memory, small chunks                                              | 139ms    | 136ms    | 143ms    | 37       | -60.49ms (x0.7)      |
| Decompress from memory, small chunks (27 MB)                                     | 44ms     | 43ms     | 67ms     | 115      | -0.24ms (x0.99)      |
| Compress to file, small chunks                                                   | 170ms    | 166ms    | 180ms    | 30       | -32.23ms (x0.84)     |
| Decompress from file, small chunks                                               | 45ms     | 44ms     | 66ms     | 112      | -1.21ms (x0.97)      |
| Calculate extra terrestrial radiation (100 years, hourly)                        | 63ms     | 62ms     | 77ms     | 80       | +16.0ms (x1.34)      |
| Interpolate radiation to 15 minutes                                              | 336ms    | 333ms    | 341ms    | 15       | +90ms (x1.37)        |
[ DEBUG ] Application shutting down (Vapor/Application.swift:136)

On the other hand I was watching closely the CPU load 8c/16T and I noticed only 2T got up to 100% while the remaining cores didn't get any load. Isn't a benchmark supposed to push the CPU to the max especially as those operation are CPU hungry. This not WSL issue 

Using for example stress tool under WSL  bring all cores to 100%.
I can do same benchmark on native Ubuntu 22.04.

So back to the cron. Then You advise to use your default.
When I install packages shoudl this be under root or current user or openmeteo-api user? I noticed on either docker or packages (Ubuntu 22.04) there were a lot of failing install due to permissions.

**Reply by patrick-zippenfenig** [reply]
Yes, HDDs cannot keep up for the conversion process. A lot of random IO is required to convert spatial to time series data.

No, the benchmark is not designed to stress test CPUs. Downloading and converting data is mostly a single core load. However, usually multiple downloader run in parallel. Only the API to serve data is designed to run on multiple cores in parallel. The benchmark is mostly there to detect fundamental performance issues. E.g. the first test is significantly slower. I suspect that Apple M1 is using different trigonometric functions that are much faster. All other tests perform as expected from a newer Ryzen CPU.

Typically the API process is running as openmeteo-api. In terms of permissions you only have to make sure the download and conversion processes are allowed to write to the data directory
